### PR TITLE
PP-834: fixed releasing queue/server level resource on suspension

### PR DESCRIPTION
**Pull request by arungrover**
#### Issue-ID

* **[PP-834](https://pbspro.atlassian.net/browse/PP-834)**

#### Problem description
Changes made for [PP-734](https://pbspro.atlassian.net/browse/PP-734) did not take into consideration the resources that might get released only on server/queue (resources with flag 'q'). It resulted into these resources not being free (after suspension) even when they were part of restrict_resource_to_release_on_suspend.

#### Cause / Analysis
resource_released_list was based off of resources_released which is made by looking at exec_vnode. This wasn't the right way to do it. It should have looked at other job's resources which are consumable but not a node level resource.

#### Solution description
Changed the way resources_released_list is being populated.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_release_limited_res_suspend.py [hunk 1]
```diff
@@ -521,3 +521,142 @@ def test_delete_res_busy_on_res_to_release_list(self):
...
+                         msg="pbs should not release resource foo")
+
+        self.assertEqual(rv[0][ras_bar], "30",
+                         msg="pbs did not release resource bar")
+
+    def test_server_res_release_upon_suspension(self):
+        """
+        Create 2 consumable resources and set it on server,
+        set one of those resouces in restrict_res_to_release_on_suspend,
+        submit a job requesting for thes resources, check if the resource
```
Comment (lisa-altair) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: typo: "thes"

File: test/tests/functional/pbs_release_limited_res_suspend.py [hunk 2]
```diff
@@ -521,3 +521,142 @@ def test_delete_res_busy_on_res_to_release_list(self):
...
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        a = {ATTR_rescavail + ".foo": '100',
+             ATTR_rescavail + ".bar": '100'}
+        self.server.manager(MGR_CMD_SET, QUEUE, a, id="workq")
+
+        j1 = Job(TEST_USER)
+        j1.set_attributes({ATTR_l+'.ncpus': '4',
+                           ATTR_l+'.foo': '30',
+                           ATTR_l+'.bar': '30'})
```
Comment (lisa-altair) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: You should probably differ the requested amount of foo and bar so it's clearer which resource you are dealing with.

File: test/tests/functional/pbs_release_limited_res_suspend.py [hunk 3]
```diff
@@ -521,3 +521,142 @@ def test_delete_res_busy_on_res_to_release_list(self):
...
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        a = {ATTR_rescavail + ".foo": '100',
+             ATTR_rescavail + ".bar": '100'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        j1 = Job(TEST_USER)
+        j1.set_attributes({ATTR_l+'.ncpus': '4',
+                           ATTR_l+'.foo': '30',
+                           ATTR_l+'.bar': '30'})
```
Comment (lisa-altair) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: I'd prefer the requested amounts for foo and bar differ.

File: test/tests/functional/pbs_release_limited_res_suspend.py [hunk 4]
```diff
@@ -521,3 +521,142 @@ def test_delete_res_busy_on_res_to_release_list(self):
...
+
+        rv = self.server.status(
+            SERVER, [ras_foo, ras_bar])
+        self.assertNotEqual(rv, None)
+
+        self.assertEqual(rv[0][ras_foo], "0",
+                         msg="pbs should not release resource foo")
+
+        self.assertEqual(rv[0][ras_bar], "30",
+                         msg="pbs did not release resource bar")
```
Comment (lisa-altair) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: Same comment about the error messages potentially needing to be switched.

File: test/tests/functional/pbs_release_limited_res_suspend.py [hunk 5]
```diff
@@ -521,3 +521,142 @@ def test_delete_res_busy_on_res_to_release_list(self):
...
+
+    def test_node_custom_res_release_upon_suspension(self):
+        """
+        Create 2 consumable resources and set it on node,
+        set one of those resouces in restrict_res_to_release_on_suspend,
+        submit a job requesting for thes resources, check if the resource
```
Comment (lisa-altair) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: typo: "thes"

File: test/tests/functional/pbs_release_limited_res_suspend.py [hunk 6]
```diff
@@ -521,3 +521,142 @@ def test_delete_res_busy_on_res_to_release_list(self):
...
+
+        rv = self.server.status(
+            NODE, [ras_foo, ras_bar], id=self.mom.shortname)
+        self.assertNotEqual(rv, None)
+
+        self.assertEqual(rv[0][ras_foo], "0",
+                         msg="pbs should not release resource foo")
+
+        self.assertEqual(rv[0][ras_bar], "30",
+                         msg="pbs did not release resource bar")
```
Comment (lisa-altair) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: Same comment about the error messages potentially needing to be switched.

File: test/tests/functional/pbs_release_limited_res_suspend.py (hunk 2 above)
Comment (arungrover) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: I'll change them

File: test/tests/functional/pbs_release_limited_res_suspend.py [hunk 7]
```diff
@@ -521,3 +521,142 @@ def test_delete_res_busy_on_res_to_release_list(self):
...
+
+        ras_foo = ATTR_rescassn + '.foo'
+        ras_bar = ATTR_rescassn + '.bar'
+
+        rv = self.server.status(
+            QUEUE, [ras_foo, ras_bar], id="workq")
+        self.assertNotEqual(rv, None)
+
+        self.assertEqual(rv[0][ras_foo], "0",
+                         msg="pbs should not release resource foo")
```
Comment (arungrover) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: You are right! error messages are messed up, I'll change them in all the places.

**Comment by arungrover** [comment]
@lisa-altair I've resolved all your review comments. Please have a look again.

File: src/server/node_manager.c [hunk 8]
```diff
@@ -8055,6 +8056,30 @@ int update_resources_rel(job *pjob, attribute *attrib, enum batch_op op)
 		} else
 			return 1;
 	}
+	/* Now iterate through all of the job resources that are present on at
+	 * queue/server level and add then to resource_release_list.
```
Comment (arungrover) [inline src/server/node_manager.c]: done

File: test/tests/functional/pbs_release_limited_res_suspend.py [hunk 9]
```diff
@@ -521,3 +521,142 @@ def test_delete_res_busy_on_res_to_release_list(self):
             self.server.manager(MGR_CMD_DELETE, RSC, id='bar')
         except PbsManagerError as e:
             self.assertTrue("Resource busy on server" in e.msg[0])
+
+    def test_queue_res_release_upon_suspension(self):
+        """
+        Create 2 consumable resources and set it on queue,
+        set one of those resouces in restrict_res_to_release_on_suspend,
+        submit a job requesting for thes resources, check if the resource
```
Comment (arungrover) [inline test/tests/functional/pbs_release_limited_res_suspend.py]: done

**Comment by arungrover** [comment]
@lisa-altair I forgot to answer your first query about using preemption for suspending a job. I think it is okay to use either one of preemption or qsig because essentially in both the cases server does the same thing of releasing resources.
If you are not too worried about it then I'll just leave it as is.

**Pull request merged**